fastpair/rust/demo/rust/src/advertisement.rs

```diff
@@ -0,0 +1,168 @@
+            distance,
+            model_id,
+            name: device_info.name().to_string(),
+            image_url: device_info.image_url().to_string(),
+        })
+    }
+
+    /// Retrieve estimated distance the BLE advertisement travelled between
+    /// the sending device and this receiver.
+    pub(crate) fn distance(&self) -> f64 {
+        self.distance
+    }
+
+    /// Retrieve the BLE Address of the advertising device.
+    pub(crate) fn address(&self) -> BleAddress {
+        self.inner.address()
+    }
+
+    /// Retrieve the Model ID advertised by this device, parsed from the
+    /// 16-bit UUID service data.
+    pub(crate) fn model_id(&self) -> &ModelId {
+        &self.model_id
+    }
+
+    pub(crate) fn name(&self) -> &String {
+        &self.name
+    }
+
+    pub(crate) fn image_url(&self) -> &String {
+        &self.image_url
+    }
+}
+
+// Implement necessary Rust constructs to easily compute the closest device.
+impl PartialEq for FpPairingAdvertisement {
+    fn eq(&self, other: &Self) -> bool {
+        self.distance == other.distance
+    }
+}
+impl Eq for FpPairingAdvertisement {}
+impl PartialOrd for FpPairingAdvertisement {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        self.distance.partial_cmp(&other.distance)
+    }
+}
+impl Ord for FpPairingAdvertisement {
+    fn cmp(&self, other: &Self) -> Ordering {
+        // `f64` isn't Ord, so we have to unwrap and hope there's only
+        // reasonable float values (e.g. NaN will panic).
+        self.distance.partial_cmp(&other.distance).unwrap()
+    }
+}
+
+/// Convert RSSI and transmit power to distance using log-distance path loss
+/// model, with reference path loss of 1m at 41dB in free space.
+/// See: https://en.wikipedia.org/wiki/Log-distance_path_loss_model.
+#[inline]
+pub(crate) fn distance_from_rssi_and_tx_power(rssi: i16, tx_power: i16) -> f64 {
+    // Source: Android Nearby implementation, `RangingUtils.java`.
+    //
+    // PL      = total path loss in db
+    // txPower = TxPower in dbm
+    // rssi    = Received signal strength in dbm
+    // PL_0    = Path loss at reference distance d_0 {@link RSSI_DROP_OFF_AT_1_M} dbm
+    // d       = length of path
+    // d_0     = reference distance  (1 m)
+    // gamma   = path loss exponent (2 in free space)
+    //
+    // Log-distance path loss (LDPL) formula:
+    //
+    // PL =    txPower - rssi                               = PL_0 + 10 * gamma  * log_10(d / d_0)
+    //         txPower - rssi               = RSSI_DROP_OFF_AT_1_M + 10 * gamma  * log_10(d / d_0)
+    //         txPower - rssi - RSSI_DROP_OFF_AT_1_M        = 10 * 2 * log_10(distanceInMeters / 1)
+    //         txPower - rssi - RSSI_DROP_OFF_AT_1_M        = 20 * log_10(distanceInMeters / 1)
+    //        (txPower - rssi - RSSI_DROP_OFF_AT_1_M) / 20  = log_10(distanceInMeters)
+    //  10 ^ ((txPower - rssi - RSSI_DROP_OFF_AT_1_M) / 20) = distanceInMeters
+
+    static RSSI_DROPOFF_AT_1_M: i16 = 41;
```

Comment:
const is probably what you want for these

Reply:
Oops I do want `const` you're right, but I'd also like `static` in the sense that they have static lifetimes and shouldn't require allocating stack space on every function call, while also not being global constants but rather being scoped to a specific method. I'll change these to `static const` for now -- but would another scheme be preferable?

Reply:
const means the compiler can inline the numbers so they never exist at runtime.